src/redisearch_rs/c_entrypoint/types_ffi/src/lib.rs

```diff
@@ -248,3 +250,143 @@ pub unsafe extern "C" fn AggregateResultIter_Free(iter: *mut RSAggregateResultIt
+///
+/// # Safety
+///
+/// The following invariants must be upheld when calling this function:
+/// - `offsets` must point to a valid [`RSOffsetVector`] and cannot be NULL.
+/// - `len` cannot be NULL and must point to an allocated memory big enough to hold an u32.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn RSOffsetVector_GetData(
+    offsets: *const RSOffsetVector,
+    len: *mut u32,
+) -> *const c_char {
+    debug_assert!(!offsets.is_null(), "offsets must not be null");
+    debug_assert!(!len.is_null(), "len must not be null");
+
+    // SAFETY: Caller is to ensure `offsets` is non-null and point to a valid RSOffsetVector.
+    let offsets = unsafe { &*offsets };
+
+    // SAFETY: Caller is to ensure `len` is non-null and point to a valid u32 memory.
+    unsafe { len.write(offsets.len) };
+    offsets.data
+}
+
+/// Set the offsets array on a [`RSOffsetVector`].
+///
+/// The [`RSOffsetVector`] will borrow the passed array so it's up to the caller to
+/// ensure it stays alive during the [`RSOffsetVector`] lifetime.
+///
+/// # Safety
+///
+/// The following invariants must be upheld when calling this function:
+/// - `offsets` must point to a valid [`RSOffsetVector`] and cannot be NULL.
+/// - `data` must point to an array of `len` offsets.
+/// - if `data` is NULL then `len` should be 0.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn RSOffsetVector_SetData(
+    offsets: *mut RSOffsetVector,
+    data: *const c_char,
+    len: u32,
+) {
+    debug_assert!(!offsets.is_null(), "offsets must not be null");
+    debug_assert!(
+        !data.is_null() || len == 0,
+        "data must not be null if len is higher than 0"
+    );
+
+    // SAFETY: Caller is to ensure `offsets` is non-null and point to a valid RSOffsetVector.
+    let offsets = unsafe { &mut *offsets };
+
+    offsets.data = data as _;
+    offsets.len = len;
+}
+
+/// Free the data inside an [`RSOffsetVector`]'s offset
+///
+/// # Safety
+///
+/// The following invariants must be upheld when calling this function:
+/// - `offsets` must point to a valid [`RSOffsetVector`] and cannot be NULL.
+/// - The data pointer of `offsets` had been allocated via the global allocator
+///   and points to an array matching the length of `offsets`.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn RSOffsetVector_FreeData(offsets: *mut RSOffsetVector) {
+    debug_assert!(!offsets.is_null(), "offsets must not be null");
+
+    // SAFETY: Caller is to ensure `offsets` is non-null and point to a valid RSOffsetVector.
+    let offsets = unsafe { &mut *offsets };
+
+    if offsets.data.is_null() {
+        return;
+    }
+
+    let layout = Layout::array::<c_char>(offsets.len as usize).unwrap();
+    // SAFETY: Caller is to ensure data has been allocated via the global allocator
+    // and points to an array matching the length of `offsets`.
+    unsafe { std::alloc::dealloc(offsets.data.cast(), layout) };
+
+    offsets.data = std::ptr::null_mut();
+    offsets.len = 0;
```

Comment:
It might be worth moving this to a method on `RSOffsetVector`. Then the next PR will just be updates on the `inverted_index/lib.rs` file

Reply:
Actually `RSOffsetVector_FreeData` should go away when switching to a `CoW` right?

Or do you mean having all those functions being simple wrapper on `RSOffsetVector` rust methods?

Reply:
And if `RSOffsetVector` is directly replaced by a `Cow` then we won't have any custom methods any more.